stock_back2back_order_proc/stock.py

```diff
@@ -0,0 +1,552 @@
+# -*- encoding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
+#    Copyright (c) 2010-2015 Elico Corp. All Rights Reserved.
+#    Author: Yannick Gouin <yannick.gouin@elico-corp.com>
+#            Alex Duan <alex.duan@elico-corp.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+from openerp.tools.translate import _
+from openerp import netsvc
+
+
+class StockLocation(orm.Model):
+    _name = 'stock.location'
+    _inherit = "stock.location"
+    _columns = {
+        'retention_mode': fields.selection(
+            [('retention', _('Retention Mode')), ('thru', _('Thru mode'))],
+            _('Retention Mode'),
+            required=True,
+            help=_("In 'Retention mode' the system wait for the\
+                  whole quantity before the stuff is processed.\n" +
+                   "In 'Thru mode' the shipped quantity is processed regardless\
+                  of the ordered quantity.")
+        ),
+    }
+    _defaults = {
+        'retention_mode': 'retention',
+    }
+
+
+class StockPicking(orm.Model):
+    _name = 'stock.picking'
+    _inherit = "stock.picking"
+
+    def get_move_chain(self, cr, uid, move_id, context=None, move_obj=False):
+        '''Recursively get the chained moves
```

Comment:
Generally speaking, I'd rather have an iterative implementation of this than a recursive implementation. This is much more efficient in Python (while loop vs. function call, you can share local variables between calls, no risk of `MaxRecursionDepthExceeded` exceptions, and Python has no tail optimisation for recursion)

In this specific case, recursion might be ok, because the chains are generally short, but this is something you may want to keep in mind.